### ISSUE TITLE
Unsupported cipher method aes-256-ctr

### ISSUE THREAD

**Issue by markus-hsk** [defuse/php-encryption#420]
At the moment i try to fix an issue on my server:
> Cipher method not supported. This is normally caused by an outdated version of OpenSSL (and/or OpenSSL compiled for FIPS compliance). Please upgrade to a newer version of OpenSSL that supports aes-256-ctr to use this library.

After some searching i found two constants in the Core class: _CIPHER_METHOD_ and _LEGACY_CIPHER_METHOD_. Is it possible to set the cipher method from outside without manipulating the original code? Did not find any possibility to do so.

_I'm using php7.2 on different OS (Ubuntu, Alpine Linux and Windows 10) with v2.2.1 of the defuse/php-encryption package, and LibreSSL 2.7.4_

**Comment by paragonie-scott** [comment]
> Is it possible to set the cipher method from outside without manipulating the original code? Did not find any possibility to do so.

No, that constant is only used in `legacyDecrypt()`.

@defuse I think we really need to do [what I did with CipherSweet](https://github.com/paragonie/ciphersweet/blob/da967be3575b6f18823546f51f75625123d35a83/src/Util.php#L16-L47) and provide an AES-256-CTR polyfill using AES-256-ECB to prevent people from doing anything dangerous.

Of course, the "correct" fix is: Install a newer version of OpenSSL that isn't compiled in FIPS mode so AES-256-CTR is available.

But I'd rather slightly increase this library's attack surface than encourage people to do dangerous things like manipulate the library's code to use an different cipher mode.

**Comment by markus-hsk** [comment]
Sorry, if this might sounds dump, but the installed version of LibreSSL is already the latest stable version 2.7.4. So, there is no possibilty to update and i cannot see any FIPS mode hint.

**Comment by paragonie-scott** [comment]
Weird. LibreSSL removes AES-256-CTR?

**Comment by paragonie-scott** [comment]
Proposed fix in #421 

@markus-hsk Can you please verify that this branch fixes your problem?

**Comment by paragonie-scott** [comment]
If you've never tested a branch before:

`composer require defuse/php-encryption:dev-aes256ctr`

Then run `test.sh`.

**Comment by markus-hsk** [comment]
Hello and good morning,
yeah, thank you. I'll try, but might take some time. I'll come back, then it's implemented and tested.

**Comment by paragonie-scott** [comment]
Can you at least verify for me that `aes-256-ecb` is available?

```php
var_dump(\in_array('aes-256-ecb', \openssl_get_cipher_methods(), true));
```

Run that and if it says `bool(true)`, then the polyfill should be successful.

**Comment by markus-hsk** [comment]
Yeah that's possible. Yes, `aes-256-ecb` is available.

For your interest, these are all available ciphers using LibreSSL 2.7.4 under Alpine Linux 3.8.1:
`
aes-128-cbc       aes-128-ecb       aes-192-cbc       aes-192-ecb
aes-256-cbc       aes-256-ecb       base64            bf
bf-cbc            bf-cfb            bf-ecb            bf-ofb
camellia-128-cbc  camellia-128-ecb  camellia-192-cbc  camellia-192-ecb
camellia-256-cbc  camellia-256-ecb  cast              cast-cbc
cast5-cbc         cast5-cfb         cast5-ecb         cast5-ofb
chacha            des               des-cbc           des-cfb
des-ecb           des-ede           des-ede-cbc       des-ede-cfb
des-ede-ofb       des-ede3          des-ede3-cbc      des-ede3-cfb
des-ede3-ofb      des-ofb           des3              desx
idea              idea-cbc          idea-cfb          idea-ecb
idea-ofb          rc2               rc2-40-cbc        rc2-64-cbc
rc2-cbc           rc2-cfb           rc2-ecb           rc2-ofb
rc4               rc4-40
`

**Comment by markus-hsk** [comment]
Sorry, that it took so long, but now i've the testing result as you wished:

```
bash-4.4# bash test.sh
gpg: /root/.gnupg/trustdb.gpg: trustdb created
gpg: error reading key: No public key
Downloading PGP Public Key...
gpg: key 4AA394086372C20A: 523 signatures not checked due to missing keys
gpg: key 4AA394086372C20A: public key "Sebastian Bergmann <sb@sebastian-bergmann.de>" imported
gpg: Total number processed: 1
gpg:               imported: 1
gpg: no ultimately trusted keys found
pub   rsa4096 2014-07-19 [SC]
      D840 6D0D 8294 7747 2937  7831 4AA3 9408 6372 C20A
uid           [ unknown] Sebastian Bergmann <sb@sebastian-bergmann.de>
uid           [ unknown] Sebastian Bergmann <sebastian@php.net>
uid           [ unknown] Sebastian Bergmann <sebastian@thephp.cc>
uid           [ unknown] Sebastian Bergmann <sebastian@phpunit.de>
uid           [ unknown] Sebastian Bergmann <sebastian.bergmann@thephp.cc>
uid           [ unknown] [jpeg image of size 40635]
sub   rsa4096 2014-07-19 [E]

Connecting to phar.phpunit.de (188.94.27.25:443)
Connecting to phar.phpunit.de (188.94.27.25:443)
phpunit.phar         100% |********************************************************************************************************************************************************************************************|  2969k  0:00:00 ETA
Connecting to phar.phpunit.de (188.94.27.25:443)
Connecting to phar.phpunit.de (188.94.27.25:443)
phpunit.phar.asc     100% |********************************************************************************************************************************************************************************************|   801   0:00:00 ETA
gpg: Signature made Thu Feb  1 05:55:32 2018 UTC
gpg:                using RSA key 4AA394086372C20A
gpg: Good signature from "Sebastian Bergmann <sb@sebastian-bergmann.de>" [unknown]
gpg:                 aka "Sebastian Bergmann <sebastian@php.net>" [unknown]
gpg:                 aka "Sebastian Bergmann <sebastian@thephp.cc>" [unknown]
gpg:                 aka "Sebastian Bergmann <sebastian@phpunit.de>" [unknown]
gpg:                 aka "Sebastian Bergmann <sebastian.bergmann@thephp.cc>" [unknown]
gpg:                 aka "[jpeg image of size 40635]" [unknown]
gpg: WARNING: This key is not certified with a trusted signature!
gpg:          There is no indication that the signature belongs to the owner.
Primary key fingerprint: D840 6D0D 8294 7747 2937  7831 4AA3 9408 6372 C20A

Begin Unit Testing
PHPUnit 5.7.27 by Sebastian Bergmann and contributors.

Error:         No code coverage driver is available

Time: 41.74 seconds, Memory: 20.00MB

OK (1378 tests, 207576 assertions)
PHPUnit 5.7.27 by Sebastian Bergmann and contributors.

Error:         No code coverage driver is available

Time: 41 seconds, Memory: 20.00MB

OK (1378 tests, 207574 assertions)
```


Sadly it's not working yet. But now we found the possible reason for our problem. When in php-cli, there are all needed cipher methods available, -ctr and -ecb also. That's the reason, why i said that ecb is available - i checked via cli. We now found out, that under cgi and not in cli, the cipher methods list is much shorter (only 39 instead of 180 under cli). Missing -ctr and -ecb (so your fix does not help me). The used php.ini of both modes is the same. It's completely unclear, why php is doing this. Maybe you have an idea about this?

**Comment by paragonie-scott** [comment]
Is the process used by the HTTP daemon using the same LibreSSL library version?

**Comment by markus-hsk** [comment]
I assume that, but cannot ensure it. Then i check the phpinfo-result under both modes, the openssl extension prints the same outputs:

CLI:
```
openssl

OpenSSL support => enabled
OpenSSL Library Version => LibreSSL 2.7.4
OpenSSL Header Version => LibreSSL 2.7.4
Openssl default config => /etc/ssl/openssl.cnf

Directive => Local Value => Master Value
openssl.cafile => no value => no value
openssl.capath => no value => no value
```

CGI:
![2018-10-19 08_30_29-phpinfo](https://user-images.githubusercontent.com/13730463/47201203-58da3880-d379-11e8-9a28-cf8850055706.png)

Checked installed packages:
```
bash-4.4# apk list *php*ssl*
php7-openssl-7.2.10-r0 x86_64 {php7} (PHP-3.0 BSD LGPL-2.0 MIT Zend) [installed]
php5-openssl-5.6.37-r0 x86_64 {php5} (PHP-3.0)

bash-4.4# apk list *libressl*
libressl-dev-2.7.4-r0 x86_64 {libressl} (custom)
libressl2.7-libcrypto-2.7.4-r0 x86_64 {libressl} (custom) [installed]
libressl-doc-2.7.4-r0 x86_64 {libressl} (custom)
libressl2.7-libtls-2.7.4-r0 x86_64 {libressl} (custom) [installed]
libressl-2.7.4-r0 x86_64 {libressl} (custom) [installed]
libressl-dbg-2.7.4-r0 x86_64 {libressl} (custom)
libressl2.7-libssl-2.7.4-r0 x86_64 {libressl} (custom) [installed]
```

At the moment i am totally confused of this.

**Comment by markus-hsk** [comment]
We tried different solutions, but could not fix the main issue. The list of ciphers is not complete, if the requests come over browser. We think, that Apache might be the troublemaker here. We also found an old bug report on php.net with a similar problem description: https://bugs.php.net/bug.php?id=65157

Because we need to get the software back working, we now made the decision to use AES-256-CBC with php_openssl natively (not using your package).

**Comment by paragonie-scott** [comment]
If you have AES-256-CBC, you also have AES-256-ECB, which means the pull request I opened *should* fix the problem with this library. There's no reason to downgrade to unauthenticated encryption.

**Comment by markus-hsk** [comment]
Nope - it's missing! As i said, under cli there are 180 cipher methods (including CTR and ECB), via browser/ apache only 39, but _not_ CTR and _not_ ECB! Thats why your solution does not work for me - i'm sorry! CBC is the only AES-256 based cipher method available via browser and cli. The reason for this is completely unclear. 

There's an apache configuration for avaiable cipher suites: https://httpd.apache.org/docs/current/ssl/ssl_howto.html#ciphersuites We checked it but to no avail. 

For the moment we need to encrypt over CBC until we find the reason behind it.

**Comment by paragonie-scott** [comment]
This is absolutely insane. I want to phone a friend.

**Comment by paragonie-scott** [comment]
Okay, I think I can turn CBC mode into ECB mode and then use it to build CTR mode.

The end result will probably be the hackiest PHP I've ever written, and I wrote a constant-time integer multiplication loop, *twice*. (Once to support 64-bit unsigned ints on 32-bit platforms.)

**Comment by paragonie-scott** [comment]
Please see #422 instead of #421. Hopefully when this is reviewed, merged, and released, you can talk the rest of your team down from the AES-CBC ledge.

**Comment by markus-hsk** [comment]
Cipher method not supported. This is normally caused by an outdated version of OpenSSL (and/or OpenSSL compiled for FIPS compliance). Please upgrade to a newer version of OpenSSL that supports aes-256-ctr to use this library.

Trace: 
```
#0 /app/www/api/vendor/defuse/php-encryption/src/Crypto.php(291): Defuse\Crypto\RuntimeTests::runtimeTest() 
#1 /app/www/api/vendor/defuse/php-encryption/src/Crypto.php(153): Defuse\Crypto\Crypto::decryptInternal('zE09E7TJ1VvXUEe...', Object(Defuse\Crypto\KeyOrPassword), false)
```

**Comment by paragonie-scott** [comment]
Oh, I forgot about the runtime test. :\

**Comment by paragonie-scott** [comment]
The  branch has been updated. I think @defuse will need to manually merge the branches, as the Code Coverage bot won't bless the unavoidable changes (difficult to test an environment in Travis CI where OpenSSL doesn't have those ciphers).

**Comment by markus-hsk** [comment]
Sorry to let you wait, but we cannot test it easily because it's only happening in a production environment. So we need to do some steps, to get your code into the application.

But now we have a result - and it's a bad one:
`PHP Fatal error:  Allowed memory size of 536870912 bytes exhausted (tried to allocate 20480 bytes) in .../defuse/php-encryption/src/Crypto.php on line 297`

This is not caused by our application, because then i use the core openssl_... methods instead, it's working properly.

**Comment by paragonie-scott** [comment]
Does https://github.com/defuse/php-encryption/pull/422/commits/a871a42d5d8a9f8839f8c23f0ea8e4f9a95b22b7 help?

**Comment by defuse** [comment]
If you're still working on this, it would be really helpful to have as-simple-as-possible instructions for reproducing the bug, e.g. "install X in a virtual machine, then run these commands." This will make sure it's not something specific to that instance of your production environment and I'll be able to figure out what's going on quickly.

**Comment by markus-hsk** [comment]
Hey!
at the moment, i am on vacation - so not able to do anything. Next week I will return back to work. I will see what i can do. Might take some time, because of over stuff to do, but i'll try to give you the requested informations. What i can say now for the meantime: you'll need Alpine Linux latest version, php 7.2 with Apache. We are using it inside a docker container.

**Comment by markus-hsk** [comment]
@defuse 
i didn't forget this, but was not yet able to check the configuration. It will take a bit more time. Maybe until next friday. 😟

**Comment by Ramapitecus** [comment]
Hey, I had this same issue. A couple of notes that may help another person that reaches this thread. 
I'm using apache/php in alpine with docker. I had the same issue mentioned above about not all ciphers being listed when calling php from the browser. 
libressl in alpine latest already includes the aes-256-ctr cipher. And it's listed for me (I did not install apache2-ssl module). 
When I restart apache `httpd -k restart` it missing, but a restart of the container fixes the issue. 

In summary, libressl package (installed separately) includes the cipher, and I don't know exactly why `httpd -k restart' causes the cipher to stop begin listed in php but a restart of the container usually fixes the issue for me.

**Comment by markus-hsk** [comment]
Hello there,
i hope you all had nice holidays. I'm really sorry that it took so much time, but we've some issues to work on at the moment. 

I created a small code base, which you can use to built up a docker container. It is showing the problem.

[openssl_cipher_bug-master-22d561c6c1e0b5bd9387822f8229441a8164e990.zip](https://github.com/defuse/php-encryption/files/2712641/openssl_cipher_bug-master-22d561c6c1e0b5bd9387822f8229441a8164e990.zip)

The container uses Alpine Linux 3.8, php7, some php-packages and LibreSSL. (You'll see them inside the docker_install.sh script).

Is this helpful, or do you need something else?

**Comment by defuse** [comment]
Thanks @markus-hsk, I haven't looked yet but that should be all I need!

**Comment by defuse** [comment]
Closing this and praying we don't actually need CTR mode implemented with CBC mode anymore.

**Comment by RicoNijeboer** [comment]
Hello @defuse ,

Do you know of any way to fix it? Because I just got the error and everywhere I look I get told to update MAMP (Which I am not using)

Some info of my machine;
```
Package version; v2.2.1
----
>>> php -v
PHP 7.3.12 (cli) (built: Nov 29 2019 20:07:49) ( NTS )
Copyright (c) 1997-2018 The PHP Group
Zend Engine v3.3.12, Copyright (c) 1998-2018 Zend Technologies
    with Zend OPcache v7.3.12, Copyright (c) 1999-2018, by Zend Technologies
---- 
>>> httpd -v
Server version: Apache/2.4.41 (Unix)
Server built:   Nov  9 2019 07:53:54
----
>>> openssl version -a
LibreSSL 2.8.3
built on: date not available
platform: information not available
options:  bn(64,64) rc4(16x,int) des(idx,cisc,16,int) blowfish(idx)
compiler: information not available
OPENSSLDIR: "/private/etc/ssl"
----
>>> var_dump(\in_array('aes-256-ecb', \openssl_get_cipher_methods(), true));
bool(true)
----
>>> var_dump(\in_array('aes-256-ctr', \openssl_get_cipher_methods(), true))
bool(true)
----
```

Also a screenshot from the phpinfo entry;

![image](https://user-images.githubusercontent.com/26306300/70984552-fd115980-20ba-11ea-8921-f585b16def2b.png)

**Comment by scottarc** [comment]
What happens if you include those `var_dump()`s inside a PHP script and execute it from a web browser?
